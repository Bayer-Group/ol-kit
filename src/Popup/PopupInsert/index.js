export { default as PopupActionGroup } from './PopupActionGroup'
export { default as PopupActionItem } from './PopupActionItem'
export { default as PopupDataList } from './PopupDataList'
export { default as PopupDefaultInsert } from './PopupDefaultInsert'
export { default as PopupDefaultPage } from './PopupDefaultPage'
export { default as PopupPageLayout } from './PopupPageLayout'
export { default as PopupPageLayoutChild } from './PopupPageLayoutChild'
export { default as PopupTabs } from './PopupTabs'
