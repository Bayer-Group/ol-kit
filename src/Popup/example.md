Popups are great for contextual content, interaction and controls that live relative to the map. The `pixel` prop determines the pixel location where the popup will be shown. By default the popup will position itself at the top left corner.

```jsx
return (
  <Popup show={true} pixel={[10, 10]} arrow={'none'} inline={true}>
    <p>This is text within a popup</p>
  </Popup>
)
```

For a better UX, use the `arrow` prop to tell the popup where it should show relative to the pixel coordinate you passed it. The default is `none` (example above) which means it doesn't show a directional arrow and positions on the top left corner. Below are some examples of the four different arrow positions you can choose from:

```jsx
return (
  <Popup show={true} pixel={[50, 80]} height={150} arrow={'left'} inline={true}>
    <p>This is <code>arrow="left"</code></p>
  </Popup>
)
```

```jsx
return (
  <Popup show={true} pixel={[350, 10]} height={150} width={500} arrow={'top'} inline={true}>
    <p>This is <code>arrow="top"</code></p>
  </Popup>
)
```

```jsx
return (
  <Popup show={true} pixel={[350, 310]} arrow={'bottom'} inline={true}>
    <p>This is <code>arrow="bottom"</code></p>
  </Popup>
)
```

```jsx
return (
  <Popup show={true} pixel={[550, 150]} width={500} arrow={'right'} inline={true}>
    <p>This is <code>arrow="right"</code></p>
  </Popup>
)
```
