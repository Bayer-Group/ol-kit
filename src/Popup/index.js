export { default as Popup } from './Popup'
export { default as PopupBase } from './PopupBase'
export * from './PopupActions'
export * from './PopupInsert'
export * from './utils'
