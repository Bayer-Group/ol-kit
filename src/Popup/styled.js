import styled from 'styled-components'

const positionContainer = (arrowDirection, [x, y], width, height) => {
  if (arrowDirection === 'top') {
    return { top: y + 16, left: x - width / 2 }
  } else if (arrowDirection === 'right') {
    return { top: y - height / 2, left: x - width - 16 }
  } else if (arrowDirection === 'bottom') {
    return { top: y - height - 16, left: x - width / 2 }
  } else if (arrowDirection === 'left') {
    return { top: y - height / 2, left: x + 16 }
  } else {
    return { top: y, left: x }
  }
}

export const Container = styled.div`
  display: ${p => p.show ? 'block' : 'none'};
  position: ${p => p.inline ? 'relative' : 'absolute'};
  background-color: #FFFFFF;
  font-family: 'ArialMT', 'Arial';
  font-weight: 400;
  z-index: 99;
  opacity: ${p => p.transparent ? 0.8 : 1};
  width: ${p => p.width}px;
  height: ${p => p.height}px;
  left: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).left}px;
  top: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).top}px;
  box-shadow: 0px 1px 5px 0px rgba(0, 0, 0, 0.2),
  0px 2px 2px 0px rgba(0, 0, 0, 0.14),
  0px 3px 1px -2px rgba(0, 0, 0, 0.12);
`

export const ArrowBox = styled.div`
  &::after, &::before {
    ${props => props.position === 'click' ? 'right' : props.position}: 100%;
    ${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &::after {
    border-${props => props.position === 'click' ? 'right' : props.position}-color: ${props => props.position === 'bottom'
  ? '#ededed'
  : '#ffffff'};
    border-width: 16px;
    margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -16px;
  }

  &::before {
    border-${props => props.position === 'click' ? 'right' : props.position}-color: rgba(0, 0, 0, 0.12);
    border-width: 17px;
    margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -17px;
  }
`
export const DragHandleIcon = styled.div`
  height: 25px;
  width: 70px;
  cursor: move;
  color: ${props => props.color || `#fff`};
  padding: 3px;
  z-index: 100;
  position: absolute;
  margin: auto;
  top: 0;
  left: 0;
  right: 0;
`
