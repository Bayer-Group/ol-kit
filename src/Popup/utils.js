import bboxTurf from '@turf/bbox'
import { lineString } from '@turf/helpers'
import centroid from '@turf/centroid'
import olMap from 'ol/Map'
import { fromLonLat } from 'ol/proj'
import GeoJSON from 'ol/format/GeoJSON'
import olLayerVector from 'ol/layer/Vector'
import olVectorTile from 'ol/layer/VectorTile'
import olSourceCluster from 'ol/source/Cluster'
import debounce from 'lodash.debounce'
import ugh from 'ugh'

/**
 * Bind multiple move listeners with the same callback
 * @function
 * @category Popup
 * @since 0.2.0
 * @param {ol.Map} map - The openlayers map to which the events are bound
 * @param {Function} callback - The callback invoked when a `change:size`, `change:resolution` or a `change:center` event was fired
 * @returns {ol.EventsKey[]} Array of openlayers event keys for unsetting listener events (use with removeMovementListener)
 */
export const addMovementListener = (map, callback) => {
  if (typeof callback !== 'function') return ugh.error('\'addMovementListener\' requires a valid openlayers map & callback function') // eslint-disable-line

  // If performance becomes an issue with catalog layers & far zoom level, these debounce levels can be adjusted
  const slowDebounce = debounce(callback, 0)
  const fastDebounce = debounce(callback, 0)

  const keys = [
    map.on('change:size', slowDebounce),
    map.getView().on('change:resolution', slowDebounce),
    map.getView().on('change:center', fastDebounce)
  ]

  return keys
}

/**
 * Remove list of event keys
 * @function
 * @category Popup
 * @since 0.2.0
 * @param {ol.Map} map - The openlayers map to which the events are bound
 * @param {Array} keys - remove the listeners via an array of event keys
 */
export const removeMovementListener = (keys = []) => {
  keys.forEach(({ target, type, listener }) => target.un(type, listener))
}

/**
 * Get all features for a given click event
 * @function
 * @category Popup
 * @since 0.2.0
 * @param {Object} event - An object with an `event` and `pixel` property
 * @param {ol.Map} event.map - The openlayers map where the layer exists
 * @param {Number[]} event.pixel - An array consisting of `x` and `y` pixel locations
 * @param {Object} [opts] - Object of optional params
 * @param {Number} [opts.hitTolerance = 3] - Additional area around features that is clickable to select them
 * @returns {Promise[]} An array of promises, each of which resolve to an object `{ layer, features }`
 */
export const getLayersAndFeaturesForEvent = (event, opts = {}) => {
  if (typeof event !== 'object') return ugh.error('getLayersAndFeaturesForEvent first arg must be an object') // eslint-disable-line
  const { map, pixel } = event
  const promises = []
  const clickCoordinate = map.getCoordinateFromPixel(pixel)

  if (!(map instanceof olMap) || !Array.isArray(pixel)) return ugh.error('getLayersAndFeaturesForEvent requires a valid openlayers map & pixel location (as an array)') // eslint-disable-line

  const setParentLayer = ({ features, layer }) => {
    return new Promise(resolve => {
      const parentLayer = layer.get('_ol_kit_parent')
      const parent = parentLayer || layer

      features.forEach((feature, i) => {
        // true makes this performant with a silent trigger: https://openlayers.org/en/latest/apidoc/module-ol_Feature.html#set
        feature.set('_ol_kit_parent', parent, true)

        return feature
      })

      resolve({ features, layer })
    })
  }

  const wfsSelector = layer => {
    // this logic only handles clicks on vector layer types
    const allowedLayerType = layer.isVectorLayer || layer instanceof olLayerVector

    // layer.getLayerState().managed is an undocumented ol prop that lets us ignore select's vector layer
    if (!layer.getLayerState().managed || !allowedLayerType) return // do nothing with these layer types

    let features = []
    const source = layer.getSource()

    if (source instanceof olSourceCluster) {
      // support for clustered feature clicks
      const clusteredFeatures = source.getClosestFeatureToCoordinate(clickCoordinate).get('features')

      features = clusteredFeatures
    } else {
      const sourceFeatures = source.getFeatures()

      sourceFeatures.forEach(sourceFeature => {
        // check if any feature on layer source is also at click location
        const isAtPixel = featuresAtPixel ? featuresAtPixel.find(f => f === sourceFeature) : null

        if (isAtPixel) features.push(sourceFeature)
      })
    }

    if (features.length) {
      const wfsPromise = setParentLayer({ features, layer })

      promises.push(wfsPromise)
    }
  }
  const wmsSelector = layer => {
    if (layer.get('_ol_kit_parent')?.isGeoserverLayer) {
      // this logic handles clicks on GeoserverLayers
      const geoserverLayer = layer.get('_ol_kit_parent')
      const coords = map.getCoordinateFromPixel(pixel)
      const wmsPromise = new Promise(async resolve => { // eslint-disable-line no-async-promise-executor
        const rawFeatures = await geoserverLayer.fetchFeaturesAtClick(coords, map)
        const { features } = await setParentLayer({ features: rawFeatures, layer })

        resolve({ features, layer })
      })

      promises.push(wmsPromise)
    }
  }

  // check for featuresAtPixel to account for hitTolerance
  const featuresAtPixel = map.getFeaturesAtPixel(pixel, {
    layerFilter: () => true,
    hitTolerance: opts.hitTolerance ? opts.hitTolerance : 3,
    checkWrapped: true
  })

  // if there's features at click, loop through the layers to find corresponding layer & features
  if (featuresAtPixel?.length) {
    map.getLayers().forEach(async layer => {
      if (layer instanceof olVectorTile) {
        const vectorTilePromise = new Promise(async resolve => { // eslint-disable-line no-async-promise-executor
          const vectorTileSourceFeatures = await layer.getSource()
            .getFeaturesInExtent(map.getView().calculateExtent(map.getSize()))
          const matchingFeaturesAtPixel = vectorTileSourceFeatures.filter(sourceFeature => {
            const { ol_uid } = sourceFeature // eslint-disable-line camelcase

            let isFeatureAtClick = false

            featuresAtPixel.forEach(feat => {
              if (feat?.ol_uid === ol_uid) isFeatureAtClick = true // eslint-disable-line camelcase
            })

            return isFeatureAtClick
          })

          const { features } = await setParentLayer({ features: matchingFeaturesAtPixel, layer })

          resolve({ features, layer })
        })

        promises.push(vectorTilePromise)
      } else {
      // handle non vector tile wfs layers
        wfsSelector(layer)
      }
    })
  }

  // this check is for wms features
  map.forEachLayerAtPixel(pixel, wmsSelector)

  return promises
}

/**
 * Get the best position for the popup to be displayed given features
 * @function
 * @category Popup
 * @param {Object} event - An object with an `event` and `pixel` property
 * @param {ol.Map} event.map - The openlayers map where the layer exists
 * @param {Number[]} event.pixel - An array consisting of `x` and `y` pixel locations
 * @param {ol.Feature[]} features - An array of features around which the popup should position
 * @param {Object} [opts]
 * @param {Number} [opts.popupHeight = 280] - The height of the popup
 * @param {Number} [opts.popupWidth = 280] - The width of the popup
 * @param {Number} [opts.arrowHeight = 16] - The height of the popup's arrow/pointer
 * @param {Number} [opts.navbarOffset = 55] - The height of the navbar
 * @param {Number[]} [opts.viewPadding = [0, 0, 0, 0]] - An array of padding to apply to the best fit logic in top, right, bottom, left order
 * @returns {Object} An object containing the arrow/pointer position, pixel location & if the popup will fit properly within the viewport
 */
export const getPopupPositionFromFeatures = (event, features = [], opts = {}) => {
  if (typeof event !== 'object' || !Array.isArray(features)) return ugh.error('getPopupPositionFromFeatures first arg must be an object & second arg array of features')
  const { map, pixel = [0, 0] } = event

  if (!(map instanceof olMap)) return ugh.error('getPopupPositionFromFeatures requires a valid openlayers map as a property of the first arg')
  if (!features.length) return { arrow: 'none', pixel, fits: false }
  const arrowHeight = opts.arrowHeight || 16
  const geoJSON = new GeoJSON({ featureProjection: 'EPSG:3857' })
  const height = opts.popupHeight || 280
  const width = opts.popupWidth || 280
  const fullHeight = height + arrowHeight
  const fullWidth = width + arrowHeight
  const [mapX, mapY] = map.getSize()

  const getPadding = (idx) => opts.viewPadding ? opts.viewPadding[idx] : calculateViewPadding(map)[idx]
  const padding = {
    top: getPadding(0),
    right: getPadding(1),
    bottom: getPadding(2),
    left: getPadding(3)
  }

  // find bbox for passed features
  const getFitsForFeatures = rawFeatures => {
    // create a new array so original features are not mutated when _ol_kit_parent is nullified
    const features = rawFeatures.map(feature => {
      const clone = feature.clone()

      // this removes a ref to _ol_kit_parent to solve circularJSON bug
      clone.set('_ol_kit_parent', null)
      clone.set('features', null)

      return clone
    })
    const jsonFeatures = geoJSON.writeFeatures(features)
    const [minX, minY, maxX, maxY] = bboxTurf(JSON.parse(jsonFeatures))

    return {
      top: getMidPixel([[minX, maxY], [maxX, maxY]]),
      right: getMidPixel([[maxX, maxY], [maxX, minY]]),
      bottom: getMidPixel([[minX, minY], [maxX, minY]]),
      left: getMidPixel([[minX, minY], [minX, maxY]])
    }
  }

  const getMidPixel = lineCoords => {
    const centerFeature = centroid(lineString(lineCoords))
    const coords = fromLonLat(centerFeature.geometry.coordinates)

    return map.getPixelFromCoordinate(coords)
  }

  const fitsRight = ([x, y]) => x + fullWidth <= mapX - padding.right && y >= (height / 2) + padding.top && y + (height / 2) <= mapY - padding.bottom // eslint-disable-line
  const fitsBelow = ([x, y]) => y + fullHeight - padding.bottom <= mapY && (width / 2) + padding.left <= x && x + (width / 2) - padding.right <= mapX // eslint-disable-line
  const fitsAbove = ([x, y]) => y - padding.top >= fullHeight && (width / 2) + padding.left <= x && x + (width / 2) - padding.right <= mapX // eslint-disable-line
  const fitsLeft = ([x, y]) => x + padding.left >= fullWidth && y >= (height / 2) + padding.top && y + (height / 2) <= mapY - padding.bottom // eslint-disable-line

  // the order of these checks determine which side is tried first (right, left, top, and then bottom)
  const getPosition = bbox => {
    if (fitsRight(bbox.right)) return { arrow: 'left', pixel: bbox.right, fits: true }
    if (fitsLeft(bbox.left)) return { arrow: 'right', pixel: bbox.left, fits: true }
    if (fitsAbove(bbox.top)) return { arrow: 'bottom', pixel: bbox.top, fits: true }
    if (fitsBelow(bbox.bottom)) return { arrow: 'top', pixel: bbox.bottom, fits: true }

    if (opts.lastPosition) {
      if (opts.lastPosition.arrow === 'left') return { arrow: 'left', pixel: bbox.right, fits: false }
      if (opts.lastPosition.arrow === 'top') return { arrow: 'top', pixel: bbox.bottom, fits: false }
      if (opts.lastPosition.arrow === 'bottom') return { arrow: 'bottom', pixel: bbox.top, fits: false }
      if (opts.lastPosition.arrow === 'right') return { arrow: 'right', pixel: bbox.left, fits: false }
      if (opts.lastPosition.arrow === 'none') return { arrow: 'none', pixel: opts.lastPosition.pixel, fits: false }
    }

    // if none of the above return, it doesn't fit on any side (it's on top of or within)
    return { arrow: 'none', pixel, fits: false }
  }

  const fitsForFeature = getFitsForFeatures(features)

  return getPosition(fitsForFeature)
}

/**
 * Calculate bounding box of elements on page with _popup_boundary class and returns padding array excluding area of these elements
 * @function
 * @category Popup
 * @param {olMap} map - An instance of an openlayers map
 * @param {Object} opts
 * @returns {Array} - Array of view padding pixel numbers: [top, right, bottom, left]
 */
export const calculateViewPadding = (map, opts = {}) => {
  if (!(map instanceof olMap)) return ugh.error('calculateViewPadding requires a valid openlayers map as arg')
  const viewPadding = [0, 0, 0, 0]
  const navbarOffset = opts.navbarOffset || 55
  const boundaryElements = Array.from(document.getElementsByClassName('_popup_boundary'))
  const [mapX, mapY] = map.getSize()

  boundaryElements.forEach(elem => {
    const bbox = elem.getBoundingClientRect()
    const isOffScreen = bbox.x < 0 || bbox.x >= mapX || bbox.y < navbarOffset || bbox.y >= mapY

    if (!isOffScreen) {
      if (bbox.right < mapX / 2) {
        // set left padding for elements on left half of map
        const newPadding = bbox.right

        if (viewPadding[3] < newPadding) viewPadding[3] = newPadding
      } else if (bbox.left > mapX / 2) {
        // set right padding for elements on right half of map
        const newPadding = mapX - bbox.left

        if (viewPadding[1] < newPadding) viewPadding[1] = newPadding
      } else if (bbox.top > mapY / 2) {
        // set bottom padding for elements on bottom half of map
        const newPadding = (mapY + navbarOffset) - bbox.top

        if (viewPadding[2] < newPadding) viewPadding[2] = newPadding
      } else if (bbox.y > navbarOffset) {
        // set top padding for elements lower than navbar
        const newPadding = bbox.bottom

        if (viewPadding[0] < newPadding) viewPadding[0] = newPadding
      }
    }
  })

  return viewPadding
}

/**
 * Remove blacklisted attributes (geom & geometry & _ol_kit*) from an object
 * @function
 * @category Popup
 * @since 0.11.0
 * @param {Object} properties - A feature attribute object
 * @returns {Object} A filtered attribute object
 */
export const sanitizeProperties = properties => {
  const blacklist = ['geom', 'geometry']
  const sanitized = {}

  for (const key in properties) {
    const keyLower = key.toLowerCase()

    if (!blacklist.includes(keyLower) && !keyLower.includes('_ol_kit')) {
      sanitized[key] = properties[key]
    }
  }

  return sanitized
}
