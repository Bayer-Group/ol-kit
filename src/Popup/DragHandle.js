import React from 'react'
import PropTypes from 'prop-types'
import { DragHandleIcon } from './styled'

export default function DragHandle (props) {
  const { color, height } = props

  return (
    <DragHandleIcon color={color} height={height} className='handle'>
      <svg width='65px' height='7px' viewBox='0 0 32 7' version='1.1' xmlns='http://www.w3.org/2000/svg' >
        <title>Drag icon</title>
        <g stroke='none' strokeWidth='1' fill='none' fillRule='evenodd'>
          <g id='2' transform='translate(-704.000000, -63.000000)' fill='#C1C1C1'>
            <g id='Bottom-Sheet' transform='translate(0.000000, 59.000000)'>
              <g id='sheet-actions' transform='translate(704.000000, 4.000000)'>
                <g id='Drag-icon'>
                  <rect id='Rectangle' x='0' y='0' width='32' height='2' rx='1'></rect>
                  <rect id='Rectangle-Copy-4' x='0' y='5' width='32' height='2' rx='1'></rect>
                </g>
              </g>
            </g>
          </g>
        </g>
      </svg>
    </DragHandleIcon>
  )
}

DragHandle.defaultProps = {
  color: 'grey',
  height: '7px'
}

DragHandle.propTypes = {
  color: PropTypes.string,
  height: PropTypes.string
}
