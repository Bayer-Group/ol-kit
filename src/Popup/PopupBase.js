import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Draggable from 'react-draggable'

import DragHandle from './DragHandle'
import { ArrowBox, Container } from './styled'

/**
 * @component
 * @category Popup
 * @since 0.2.0
 */
class PopupBase extends Component {
  state = {
    dragged: false,
    pinnedPixel: this.props.pixel,
    transparent: false
  }

  onStart = () => {
    const { arrow, onPopupDragStart, pixel, transparentDuringDrag } = this.props
    // this hard sets the arrow direction at the time of initial drag
    const lastArrow = !this.state.dragged ? arrow : this.state.lastArrow
    const pinnedPixel = !this.state.dragged ? pixel : this.state.pinnedPixel

    this.setState({
      dragged: true,
      lastArrow,
      pinnedPixel,
      transparent: transparentDuringDrag
    })
    onPopupDragStart()
  }

  onStop = e => {
    const { pinnedPixel } = this.state

    this.setState({
      transparent: false
    })
    this.props.onPopupDragEnd({ ...e, pinnedPixel })
  }

  handleDrag = e => {
    this.props.onPopupDrag(e)
  }

  render () {
    const { arrow: arrowProp, children, draggable, height, inline, pixel: pixelProp, show, width } = this.props
    const { dragged, lastArrow, pinnedPixel, transparent } = this.state
    const arrowTranslator = {
      top: 'bottom',
      bottom: 'top',
      left: 'right',
      right: 'left',
      none: 'none'
    }
    const pixel = dragged ? pinnedPixel : pixelProp
    const arrow = dragged ? lastArrow : arrowProp

    return (
      <Draggable
        axis='both'
        bounds='parent'
        handle='.handle'
        onDrag={this.handleDrag}
        onStart={this.onStart}
        onStop={this.onStop}>
        <Container
          arrow={arrow}
          height={height}
          inline={inline}
          pixel={pixel}
          show={show}
          transparent={transparent}
          width={width}>
          {draggable ? <DragHandle /> : null}
          {!dragged && <ArrowBox position={arrowTranslator[arrow]} />}
          {children}
        </Container>
      </Draggable>
    )
  }
}

PopupBase.defaultProps = {
  arrow: 'none',
  draggable: true,
  height: 280,
  inline: false,
  onPopupDrag: () => {},
  onPopupDragEnd: () => {},
  onPopupDragStart: () => {},
  pixel: [0, 0],
  show: false,
  transparentDuringDrag: true,
  width: 280
}

PopupBase.propTypes = {
  /** The position of the popup's arrow (`top`, `right`, `bottom`, `left` or `none`) */
  arrow: PropTypes.string,
  /** The content of the popup */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  draggable: PropTypes.bool,
  /** The height of the popup */
  height: PropTypes.number,
  /** Render the component inline (without absolute positioning) */
  inline: PropTypes.bool,
  /** Callback fired during drags when draggable is true */
  onPopupDrag: PropTypes.func,
  /** Callback fired after a drag when draggable is true */
  onPopupDragEnd: PropTypes.func,
  /** Callback fired at the beginning of a drag when draggable is true */
  onPopupDragStart: PropTypes.func,
  /** The pixel coordinate where the popup should render */
  pixel: PropTypes.array,
  /** Show/hide the popup */
  show: PropTypes.bool,
  /** Set PopupBase to see-through during a drag when draggable is true */
  transparentDuringDrag: PropTypes.bool,
  /** The width of the popup */
  width: PropTypes.number
}

export default PopupBase
