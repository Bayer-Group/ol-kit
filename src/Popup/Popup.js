import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import olInteractionDraw from 'ol/interaction/draw'

import { connectToMap } from 'Map'
import PopupBase from './PopupBase'
import PopupDefaultInsert from './PopupInsert/PopupDefaultInsert'
import { addMovementListener, getLayersAndFeaturesForEvent, getPopupPositionFromFeatures, removeMovementListener } from './utils'

/**
 * @component
 * @category Popup
 * @since 0.2.0
 */
class Popup extends Component {
  constructor (props) {
    super(props)

    this.state = {
      clickCoordinate: [0, 0],
      clickPixel: [0, 0],
      features: [],
      loading: false,
      popupPosition: {
        arrow: 'none',
        fits: false,
        pixel: [0, 0]
      },
      show: false
    }

    this.timer = 0
    this.isDoubleClick = false
    this.defaultState = this.state
  }

  componentDidMount () {
    const { map } = this.props

    // bind to map click events
    map.on('click', this.mapClickHandler)
    map.on('dblclick', this.mapDoubleClickHandler)
  }

  componentWillUnmount () {
    const { map } = this.props

    map.un('click', this.mapClickHandler)
    map.un('dblclick', this.mapDoubleClickHandler)
  }

  mapDoubleClickHandler = () => {
    clearTimeout(this.timer)
    this.isDoubleClick = true
    this.hidePopup()
  }

  mapClickHandler = e => {
    this.timer = setTimeout(() => {
      if (!this.isDoubleClick) {
        // Get the interactions from the map as an array.
        const interactions = e.map.getInteractions().getArray()

        // This checks to see if there is an active Draw interaction on the map and prevents the popup showing if it returns true.
        if (interactions.find((i) => i instanceof olInteractionDraw && i.get('active'))) return this.hidePopup()
        this.checkForFeaturesAtClick(e)
      }
      this.isDoubleClick = false
    }, 300)
  }

  checkForFeaturesAtClick = e => {
    const { map } = this.props
    /**
      this util returns an array of promises for each layer at the click location that resolve once wms features
      are converted to wfs and adds them to the map (immediately resolves for features already on the map)
    */
    const promises = getLayersAndFeaturesForEvent(e)

    // cancel if no layers at click location
    // do not add map movement listener or get features from layers
    if (!promises.length) return this.hidePopup()

    // when the map is panned, we need to re-calculate the position of the popup
    const popupMoveHandler = () => {
      const { show, features, popupPosition: lastPosition } = this.state
      const opts = { lastPosition } // use current position as lastPosition for animation when moving map

      // only compute new positions if the popup is showing
      if (show) {
        // this util returns a position object that is used by the popup component in render()
        const popupPosition = getPopupPositionFromFeatures({ map }, features, opts)

        this.setState({ popupPosition })
      }
    }

    // this adds three listeners (map move, resize, zoom), all of which should re-position the popup
    this.movementListener = addMovementListener(map, popupMoveHandler)
    // parse promises for layers and features
    this.getNewFeatures(e, promises)
  }

  getNewFeatures = async (e, promises) => {
    const { onMapClick } = this.props
    const popupPositionWhileLoading = getPopupPositionFromFeatures(e)

    // show popup in loading state while before resolving
    this.setState({
      clickCoordinate: e.coordinate,
      clickPixel: e.pixel,
      loading: true,
      popupPosition: popupPositionWhileLoading,
      show: true
    }, () => onMapClick(this.state))

    const layers = await Promise.all(promises)
    const parsedFeatures = layers.reduce((acc, { features }) => {
      return [...acc, ...features]
    }, [])

    // ol returns these in reverse z-index order
    const features = parsedFeatures.reverse()
    const popupPosition = getPopupPositionFromFeatures(e, features)

    this.setState({ features, loading: false, popupPosition }, () => onMapClick(this.state))
  }

  hidePopup = () => {
    const { onMapClick } = this.props

    // stop tracking movement when popup show is set to false
    this.movementListener && removeMovementListener(this.movementListener)

    this.setState({ ...this.defaultState }, () => onMapClick(this.state))
  }

  onDragEnd = e => {
    // if drag occurs in PopupBase, update pixel in state here
    this.setState({
      popupPosition: {
        ...this.state.popupPosition,
        pixel: e.pinnedPixel
      }
    })
  }

  render () {
    const { actions, children, map, show: propShow } = this.props
    const { features, loading, popupPosition: { arrow, pixel }, show: stateShow } = this.state
    const show = typeof propShow === 'boolean' ? propShow : stateShow // keep show prop as source of truth over state

    return (
      !show
        ? null
        : (
          ReactDOM.createPortal(
            <PopupBase arrow={arrow} onPopupDragEnd={this.onDragEnd} pixel={pixel} {...this.props} show={show}>
              {children || ( // default ui if no children are passed
                <PopupDefaultInsert
                  actions={actions}
                  features={features}
                  loading={loading}
                  onClose={this.hidePopup} />
              )}
            </PopupBase>,
            map.getTargetElement()
          )
        )
    )
  }
}

Popup.defaultProps = {
  onMapClick: () => {},
  show: undefined
}

Popup.propTypes = {
  /** components passed to PopupDefaultInsert to render as actions */
  actions: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** Pass components as children of Popup component */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** a reference to openlayers map object */
  map: PropTypes.object.isRequired,
  /** callback fired on map clicks with state object:
    {
      clickCoordinate: [0, 0],
      clickPixel: [0, 0],
      features: [],
      loading: false, // true after click before layers/features load
      popupPosition: {
        arrow: 'none',
        fits: true, // does the popup have room to render around the feature bbox
        pixel: [0, 0]
      },
      show: false // suggestion to display or hide popup -- <Popup show={sourceOfTruth}> (show prop takes priority)
    }
  */
  onMapClick: PropTypes.func,
  /** boolean that is respected over internal state */
  show: PropTypes.bool
}

export default connectToMap(Popup)
