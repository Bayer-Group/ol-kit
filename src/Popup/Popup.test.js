import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { prettyDOM } from '@testing-library/dom'
import olFeature from 'ol/Feature'
import olVectorLayer from 'ol/layer/Vector'
import olPoint from 'ol/geom/Point'
import olVectorSource from 'ol/source/Vector'
import { Map } from 'Map'
import { Popup } from 'Popup'

describe('<Popup />', () => {
  it('should fire onMapClick for empty map click', async () => {
    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })
    const onMapClick = jest.fn()
    const { getByTestId } = render(<Map onMapInit={onMapInit}><Popup onMapClick={onMapClick} show /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    // click the map anywhere (there are no features to trigger the popup)
    testMap.dispatchEvent({ type: 'click', map: testMap, pixel: [20, 20] })

    const hidePopupEvent = {
      clickCoordinate: [0, 0],
      clickPixel: [0, 0],
      features: [],
      loading: false,
      popupPosition: { arrow: 'none', pixel: [0, 0], fits: false },
      show: false
    }

    setTimeout(() => expect(onMapClick).toHaveBeenCalledWith(hidePopupEvent), 300)

    const defaultInsert = await getByTestId('popup-insert-default')

    expect(defaultInsert).toBeTruthy()
  })

  it.skip('should fire onMapClick for feature clicks on map', async () => {
    // set dimensions for body
    document.body.style.setProperty('width', '600px')
    document.body.style.setProperty('height', '400px')

    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })
    const onMapClick = jest.fn()
    const { getByTestId } = render(<Map onMapInit={onMapInit} fullScreen><Popup onMapClick={onMapClick} show /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    // set dimensions for map
    testMap.getTargetElement().style.setProperty('width', '600px')
    testMap.getTargetElement().style.setProperty('height', '400px')
    testMap.setSize(600, 400)
    const [minx, miny, maxx, maxy] = testMap.getView().calculateExtent([600, 400]) // eslint-disable-line no-unused-vars

    const coords = [minx, maxy]

    // add a feature to that map at a known pixel location
    const features = [new olFeature(new olPoint([coords]))]

    const source = new olVectorSource({ features })
    const vectorLayer = new olVectorLayer({ source })

    testMap.addLayer(vectorLayer)

    // click the map anywhere (there are no features to trigger the popup)
    testMap.dispatchEvent({ type: 'click', map: testMap, pixel: [0, 0] })

    const hidePopupEvent = {
      features, // should have an array with single feature above
      loading: false,
      popupPosition: { arrow: 'none', pixel: [0, 0], fits: false },
      show: true
    }

    expect(onMapClick).toHaveBeenCalledWith(hidePopupEvent)

    const defaultInsert = await getByTestId('popup-insert-default')

    expect(defaultInsert).toBeTruthy()
  })

  it('should render custom child', async () => {
    const customChild = <div data-testid='popup-insert-custom'>this is a custom popup insert</div>
    const { container, getByTestId } = render(<Map><Popup show>{customChild}</Popup></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })
    const child = await getByTestId('popup-insert-custom')

    expect(prettyDOM(child)).toMatchSnapshot()
  })
})
