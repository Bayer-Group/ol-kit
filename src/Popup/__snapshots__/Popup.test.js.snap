// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> should render custom child 1`] = `
"[36m<div[39m
  [33mdata-testid[39m=[32m\\"popup-insert-custom\\"[39m
[36m>[39m
  [0mthis is a custom popup insert[0m
[36m</div>[39m"
`;
